File: python/ui/nicegui_demo/kline_chart.js
/**
 * K线图表Vue组件 - 包装klinecharts库
 */

export default {
  template: `
    <div ref="chartContainer" :style="{ height: height }"></div>
  `,
  
  props: {
    symbol: {
      type: String,
      default: 'AAPL'
    },
    period: {
      type: String,
      default: 'day'
    },
    theme: {
      type: String,
      default: 'dark'
    },
    height: {
      type: String,
      default: '600px'
    },
    data: {
      type: Array,
      default: () => []
    }
  },
  
  data() {
    return {
      chart: null,
      currentSymbol: this.symbol,
      currentPeriod: this.period,
      currentTheme: this.theme
    };
  },
  
  watch: {
    symbol(newVal) {
      if (newVal !== this.currentSymbol) {
        this.currentSymbol = newVal;
        this.$emit('symbol-change', { symbol: newVal });
        this.requestData();
      }
    },
    
    period(newVal) {
      if (newVal !== this.currentPeriod) {
        this.currentPeriod = newVal;
        this.$emit('period-change', { period: newVal });
        this.requestData();
      }
    },
    
    theme(newVal) {
      if (newVal !== this.currentTheme) {
        this.currentTheme = newVal;
        this.updateTheme();
      }
    },
    
    data: {
      handler(newVal) {
        if (this.chart && newVal.length > 0) {
          this.chart.applyNewData(newVal);
        }
      },
      deep: true
    }
  },
  
  mounted() {
    this.initChart();
    this.requestData();
    
    // 监听窗口大小变化
    window.addEventListener('resize', this.handleResize);
  },
  
  beforeUnmount() {
    window.removeEventListener('resize', this.handleResize);
    if (this.chart) {
      this.chart.dispose();
    }
  },
  
  methods: {
    initChart() {
      // 动态加载klinecharts库
      this.loadKLineCharts().then(() => {
        if (window.klinecharts) {
          this.chart = window.klinecharts.init(this.$refs.chartContainer);
          this.updateTheme();
          
          // 添加默认指标
          this.chart.createIndicator('MA', true, { id: 'candle_pane' });
          this.chart.createIndicator('VOL');
          
          // 订阅图表事件
          this.chart.subscribeAction('onTooltipIconClick', (data) => {
            console.log('Tooltip icon clicked:', data);
          });
        }
      }).catch(error => {
        console.error('Failed to load klinecharts:', error);
      });
    },
    
    loadKLineCharts() {
      return new Promise((resolve, reject) => {
        if (window.klinecharts) {
          resolve();
          return;
        }
        
        const script = document.createElement('script');
        script.src = 'https://cdn.jsdelivr.net/npm/klinecharts/dist/klinecharts.min.js';
        script.onload = resolve;
        script.onerror = reject;
        document.head.appendChild(script);
      });
    },
    
    updateTheme() {
      if (!this.chart) return;
      
      const themeStyles = this.currentTheme === 'dark' ? {
        grid: {
          show: true,
          horizontal: {
            show: true,
            size: 1,
            color: '#292929',
            style: 'dashed'
          },
          vertical: {
            show: true,
            size: 1,
            color: '#292929',
            style: 'dashed'
          }
        },
        candle: {
          type: 'area',
          tooltip: {
            showRule: 'always',
            showType: 'standard',
            labels: ['时间: ', '开: ', '收: ', '低: ', '高: ', '成交量: '],
            text: {
              size: 12,
              color: '#D9D9D9'
            }
          },
          area: {
            lineColor: '#1677FF',
            backgroundColor: [{ offset: 0, color: 'rgba(22, 119, 255, 0.01)' }, { offset: 1, color: 'rgba(22, 119, 255, 0.2)' }]
          },
          bar: {
            upColor: '#26A69A',
            downColor: '#EF5350',
            noChangeColor: '#888888'
          }
        },
        indicator: {
          tooltip: {
            showRule: 'always',
            showType: 'standard',
            labels: ['MA5: ', 'MA10: ', 'MA30: ', 'MA60: '],
            text: {
              size: 12,
              color: '#D9D9D9'
            }
          }
        },
        xAxis: {
          axisLine: {
            show: true,
            color: '#292929'
          },
          tickLine: {
            show: true,
            length: 5,
            color: '#292929'
          },
          tickText: {
            show: true,
            color: '#848E9C',
            size: 12
          }
        },
        yAxis: {
          axisLine: {
            show: true,
            color: '#292929'
          },
          tickLine: {
            show: true,
            length: 5,
            color: '#292929'
          },
          tickText: {
            show: true,
            color: '#848E9C',
            size: 12
          }
        },
        crosshair: {
          show: true,
          horizontal: {
            show: true,
            line: {
              show: true,
              style: 'dashed',
              dashValue: [4, 2],
              size: 1,
              color: '#888888'
            },
            text: {
              show: true,
              color: '#D9D9D9',
              size: 12,
              backgroundColor: '#505050'
            }
          },
          vertical: {
            show: true,
            line: {
              show: true,
              style: 'dashed',
              dashValue: [4, 2],
              size: 1,
              color: '#888888'
            },
            text: {
              show: true,
              color: '#D9D9D9',
              size: 12,
              backgroundColor: '#505050'
            }
          }
        }
      } : {
        grid: {
          show: true,
          horizontal: {
            show: true,
            size: 1,
            color: '#F0F0F0',
            style: 'dashed'
          },
          vertical: {
            show: true,
            size: 1,
            color: '#F0F0F0',
            style: 'dashed'
          }
        },
        candle: {
          type: 'area',
          tooltip: {
            showRule: 'always',
            showType: 'standard',
            labels: ['时间: ', '开: ', '收: ', '低: ', '高: ', '成交量: '],
            text: {
              size: 12,
              color: '#333333'
            }
          },
          area: {
            lineColor: '#1677FF',
            backgroundColor: [{ offset: 0, color: 'rgba(22, 119, 255, 0.01)' }, { offset: 1, color: 'rgba(22, 119, 255, 0.2)' }]
          },
          bar: {
            upColor: '#26A69A',
            downColor: '#EF5350',
            noChangeColor: '#888888'
          }
        },
        indicator: {
          tooltip: {
            showRule: 'always',
            showType: 'standard',
            labels: ['MA5: ', 'MA10: ', 'MA30: ', 'MA60: '],
            text: {
              size: 12,
              color: '#333333'
            }
          }
        },
        xAxis: {
          axisLine: {
            show: true,
            color: '#F0F0F0'
          },
          tickLine: {
            show: true,
            length: 5,
            color: '#F0F0F0'
          },
          tickText: {
            show: true,
            color: '#666666',
            size: 12
          }
        },
        yAxis: {
          axisLine: {
            show: true,
            color: '#F0F0F0'
          },
          tickLine: {
            show: true,
            length: 5,
            color: '#F0F0F0'
          },
          tickText: {
            show: true,
            color: '#666666',
            size: 12
          }
        },
        crosshair: {
          show: true,
          horizontal: {
            show: true,
            line: {
              show: true,
              style: 'dashed',
              dashValue: [4, 2],
              size: 1,
              color: '#888888'
            },
            text: {
              show: true,
              color: '#333333',
              size: 12,
              backgroundColor: '#FFFFFF'
            }
          },
          vertical: {
            show: true,
            line: {
              show: true,
              style: 'dashed',
              dashValue: [4, 2],
              size: 1,
              color: '#888888'
            },
            text: {
              show: true,
              color: '#333333',
              size: 12,
              backgroundColor: '#FFFFFF'
            }
          }
        }
      };
      
      this.chart.setStyles(themeStyles);
    },
    
    requestData() {
      this.$emit('data-request', {
        symbol: this.currentSymbol,
        period: this.currentPeriod
      });
    },
    
    handleResize() {
      if (this.chart) {
        this.chart.resize();
      }
    },
    
    addIndicator(indicator) {
      if (this.chart) {
        this.chart.createIndicator(indicator, true);
      }
    },
    
    removeIndicator(indicator) {
      if (this.chart) {
        this.chart.removeIndicator(indicator);
      }
    },
    
    resize() {
      this.handleResize();
    }
  }
};

File: python/ui/nicegui_demo/kline_chart.py
"""
K线图表自定义组件 - 包装klinecharts库
"""

from nicegui.element import Element
from nicegui.events import GenericEventArguments, handle_event
from typing import Optional, Dict, List, Callable, Any


class KLineChart(Element, component='kline_chart.js'):
    """
    K线图表组件，基于klinecharts库
    
    支持功能：
    - K线图表渲染
    - 多周期切换
    - 技术指标显示
    - 主题切换
    - 数据实时更新
    """
    
    def __init__(
        self,
        symbol: str = 'AAPL',
        period: str = 'day',
        theme: str = 'dark',
        height: str = '600px'
    ) -> None:
        super().__init__()
        self._props['symbol'] = symbol
        self._props['period'] = period
        self._props['theme'] = theme
        self._props['height'] = height
        self._props['data'] = []
        
        # 注册事件处理器
        self.on('period-change', self._handle_period_change)
        self.on('symbol-change', self._handle_symbol_change)
        self.on('data-request', self._handle_data_request)
    
    def _handle_period_change(self, e: GenericEventArguments) -> None:
        """处理周期变更事件"""
        if hasattr(self, '_period_callback'):
            self._period_callback(e.args)
    
    def _handle_symbol_change(self, e: GenericEventArguments) -> None:
        """处理股票代码变更事件"""
        if hasattr(self, '_symbol_callback'):
            self._symbol_callback(e.args)
    
    def _handle_data_request(self, e: GenericEventArguments) -> None:
        """处理数据请求事件"""
        if hasattr(self, '_data_callback'):
            self._data_callback(e.args)
    
    def on_period_change(self, callback: Callable[[Dict], None]) -> 'KLineChart':
        """注册周期变更回调"""
        self._period_callback = callback
        return self
    
    def on_symbol_change(self, callback: Callable[[Dict], None]) -> 'KLineChart':
        """注册股票代码变更回调"""
        self._symbol_callback = callback
        return self
    
    def on_data_request(self, callback: Callable[[Dict], None]) -> 'KLineChart':
        """注册数据请求回调"""
        self._data_callback = callback
        return self
    
    def set_data(self, data: List[Dict[str, Any]]) -> None:
        """设置K线数据"""
        self._props['data'] = data
        self.update()
    
    def set_theme(self, theme: str) -> None:
        """设置主题"""
        self._props['theme'] = theme
        self.update()
    
    def set_period(self, period: str) -> None:
        """设置周期"""
        self._props['period'] = period
        self.update()
    
    def set_symbol(self, symbol: str) -> None:
        """设置股票代码"""
        self._props['symbol'] = symbol
        self.update()
    
    def add_indicator(self, indicator: str) -> None:
        """添加技术指标"""
        self.run_method('addIndicator', indicator)
    
    def remove_indicator(self, indicator: str) -> None:
        """移除技术指标"""
        self.run_method('removeIndicator', indicator)
    
    def resize(self) -> None:
        """调整图表大小"""
        self.run_method('resize')
